File: netlify/functions/_shared/group-manifest.ts
export const GROUP_MANIFEST_TYPE = "omgskills.skill_group" as const;
export const GROUP_MANIFEST_VERSION = 2 as const;

export const groupManifestMetadataOnlyReasons = [
  "release_unavailable",
  "source_unavailable",
  "source_mismatch",
  "incomplete_release",
  "invalid_release",
  "release_source_mismatch",
  "synced_missing",
  "synced_local_only",
  "synced_ambiguous",
  "synced_unresolved"
] as const;

export type GroupManifestMetadataOnlyReason =
  (typeof groupManifestMetadataOnlyReasons)[number];

export type GroupManifestSourceKind = "catalog" | "public_github" | "private_github";

export type GroupManifestSourceInput = {
  id: string;
  kind: GroupManifestSourceKind | string;
  normalizedRoot?: string | null;
  catalogSkillId?: string | null;
  repositoryId?: string | null;
  repositorySlug?: string | null;
  tombstonedAt?: Date | string | null;
};

export type GroupManifestReleaseInput = {
  id?: string | null;
  sourceId?: string | null;
  commitSha?: string | null;
  treeSha?: string | null;
  skillMdSha?: string | null;
};

export type GroupManifestSyncedIdentityInput = {
  identityStatus?: "resolved" | "ambiguous" | "localOnly" | string | null;
  catalogSkillId?: string | null;
  isCurrent?: boolean | null;
};

export type GroupManifestItemInput = {
  id: string;
  kind: "catalog" | "github" | "synced";
  position: number;
  name: string;
  description?: string | null;
  note?: string | null;
  catalogSkillId?: string | null;
  metadataOnlyReason?: string | null;
  source?: GroupManifestSourceInput | null;
  release?: GroupManifestReleaseInput | null;
  syncedIdentity?: GroupManifestSyncedIdentityInput | null;
};

export type GroupManifestInput = {
  group: {
    id: string;
    name: string;
    description?: string | null;
    slug: string;
    revision: number;
  };
  items: GroupManifestItemInput[];
};

type ManifestRelease = {
  id: string;
  commitSha: string;
  treeSha: string;
  skillMdSha: string;
};

type ManifestSource =
  | {
      id: string;
      kind: "catalog";
      catalogSkillId: string;
      normalizedRoot: string;
    }
  | {
      id: string;
      kind: "public_github";
      repositoryId: string;
      repositorySlug: string;
      normalizedRoot: string;
    }
  | {
      id: string;
      kind: "private_github";
    };

export type GroupManifestItem = {
  id: string;
  kind: GroupManifestItemInput["kind"];
  position: number;
  name: string;
  description: string | null;
  note: string | null;
  installability:
    | {
        status: "installable";
        source: ManifestSource;
        release: ManifestRelease;
      }
    | {
        status: "metadata_only";
        reason: GroupManifestMetadataOnlyReason;
      };
};

export type GroupManifest = {
  type: typeof GROUP_MANIFEST_TYPE;
  version: typeof GROUP_MANIFEST_VERSION;
  group: {
    id: string;
    name: string;
    description: string | null;
    slug: string;
    revision: number;
  };
  items: GroupManifestItem[];
};

const shaPattern = /^[0-9a-f]{40}$/;
const metadataOnlyReasonSet = new Set<string>(groupManifestMetadataOnlyReasons);

function requireText(value: unknown, label: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${label} is required`);
  }
  return value.trim();
}

function optionalText(value: unknown): string | null {
  return typeof value === "string" && value.trim() ? value.trim() : null;
}

function requireItemKind(value: unknown): GroupManifestItemInput["kind"] {
  if (value !== "catalog" && value !== "github" && value !== "synced") {
    throw new Error("item kind must be catalog, github, or synced");
  }
  return value;
}

function explicitMetadataOnlyReason(
  value: unknown
): GroupManifestMetadataOnlyReason | null {
  return typeof value === "string" && metadataOnlyReasonSet.has(value)
    ? value as GroupManifestMetadataOnlyReason
    : null;
}

function syncedMetadataOnlyReason(
  identity: GroupManifestSyncedIdentityInput | null | undefined
): GroupManifestMetadataOnlyReason {
  if (!identity || identity.isCurrent === false) {
    return "synced_missing";
  }
  if (identity.identityStatus === "localOnly") {
    return "synced_local_only";
  }
  if (identity.identityStatus === "ambiguous") {
    return "synced_ambiguous";
  }
  if (identity.identityStatus !== "resolved" || !optionalText(identity.catalogSkillId)) {
    return "synced_unresolved";
  }
  return "release_unavailable";
}

function metadataOnly(
  reason: GroupManifestMetadataOnlyReason
): GroupManifestItem["installability"] {
  return { status: "metadata_only", reason };
}

function normalizeInstallability(
  item: GroupManifestItemInput
): GroupManifestItem["installability"] {
  const explicitReason = explicitMetadataOnlyReason(item.metadataOnlyReason);
  if (explicitReason) {
    return metadataOnly(explicitReason);
  }

  const source = item.source;
  const release = item.release;
  if (!source && !release) {
    return metadataOnly(
      item.kind === "synced"
        ? syncedMetadataOnlyReason(item.syncedIdentity)
        : "release_unavailable"
    );
  }
  if (!source || source.tombstonedAt) {
    return metadataOnly("source_unavailable");
  }
  if (!release) {
    return metadataOnly(
      item.kind === "synced"
        ? syncedMetadataOnlyReason(item.syncedIdentity)
        : "release_unavailable"
    );
  }

  const releaseId = optionalText(release.id);
  const releaseSourceId = optionalText(release.sourceId);
  const commitSha = optionalText(release.commitSha);
  const treeSha = optionalText(release.treeSha);
  const skillMdSha = optionalText(release.skillMdSha);
  if (!releaseId || !releaseSourceId || !commitSha || !treeSha || !skillMdSha) {
    return metadataOnly("incomplete_release");
  }
  if (
    !shaPattern.test(commitSha)
    || !shaPattern.test(treeSha)
    || !shaPattern.test(skillMdSha)
  ) {
    return metadataOnly("invalid_release");
  }
  if (releaseSourceId !== source.id) {
    return metadataOnly("release_source_mismatch");
  }

  const sourceId = optionalText(source.id);
  const normalizedRoot = optionalText(source.normalizedRoot);
  if (!sourceId) {
    return metadataOnly("source_unavailable");
  }

  let manifestSource: ManifestSource;
  if (source.kind === "catalog") {
    const catalogSkillId = optionalText(source.catalogSkillId);
    const expectedCatalogSkillId = item.kind === "synced"
      ? optionalText(item.syncedIdentity?.catalogSkillId)
      : optionalText(item.catalogSkillId);
    if (
      !catalogSkillId
      || !normalizedRoot
      || !expectedCatalogSkillId
      || expectedCatalogSkillId.toLowerCase() !== catalogSkillId.toLowerCase()
    ) {
      return metadataOnly("source_mismatch");
    }
    manifestSource = {
      id: sourceId,
      kind: "catalog",
      catalogSkillId,
      normalizedRoot
    };
  } else if (source.kind === "public_github") {
    if (item.kind === "catalog") {
      return metadataOnly("source_mismatch");
    }
    const repositoryId = optionalText(source.repositoryId);
    const repositorySlug = optionalText(source.repositorySlug);
    if (!repositoryId || !repositorySlug || !normalizedRoot) {
      return metadataOnly("source_mismatch");
    }
    manifestSource = {
      id: sourceId,
      kind: "public_github",
      repositoryId,
      repositorySlug,
      normalizedRoot
    };
  } else if (source.kind === "private_github") {
    if (item.kind === "catalog") {
      return metadataOnly("source_mismatch");
    }
    const repositoryId = optionalText(source.repositoryId);
    const repositorySlug = optionalText(source.repositorySlug);
    if (!repositoryId || !repositorySlug || !normalizedRoot) {
      return metadataOnly("source_mismatch");
    }
    manifestSource = { id: sourceId, kind: "private_github" };
  } else {
    return metadataOnly("source_mismatch");
  }

  return {
    status: "installable",
    source: manifestSource,
    release: {
      id: releaseId,
      commitSha,
      treeSha,
      skillMdSha
    }
  };
}

export function buildGroupManifest(input: GroupManifestInput): GroupManifest {
  const revision = input.group.revision;
  if (!Number.isInteger(revision) || revision < 1) {
    throw new Error("group revision must be a positive integer");
  }

  const items = [...input.items]
    .sort((left, right) => left.position - right.position || left.id.localeCompare(right.id))
    .map((item, position): GroupManifestItem => {
      if (!Number.isInteger(item.position) || item.position < 0) {
        throw new Error("item position must be a non-negative integer");
      }
      return {
        id: requireText(item.id, "item id"),
        kind: requireItemKind(item.kind),
        position,
        name: requireText(item.name, "item name"),
        description: optionalText(item.description),
        note: optionalText(item.note),
        installability: normalizeInstallability(item)
      };
    });

  return {
    type: GROUP_MANIFEST_TYPE,
    version: GROUP_MANIFEST_VERSION,
    group: {
      id: requireText(input.group.id, "group id"),
      name: requireText(input.group.name, "group name"),
      description: optionalText(input.group.description),
      slug: requireText(input.group.slug, "group slug"),
      revision,
    },
    items
  };
}

export function serializeGroupManifest(input: GroupManifestInput): string {
  return `${JSON.stringify(buildGroupManifest(input), null, 2)}\n`;
}

File: menubar/Sources/omgskills/GroupManifest.swift
import Foundation

enum GroupManifestValidationError: Error, Equatable, Sendable {
    case unexpectedType(String)
    case unsupportedVersion(Int)
    case invalidField(String)
    case tooManyItems(Int)
    case duplicateItemID(String)
    case duplicatePosition(Int)
    case invalidPosition(expected: Int, actual: Int)
}

struct GroupManifest: Decodable, Equatable, Sendable {
    static let expectedType = "omgskills.skill_group"
    static let supportedVersion = 2
    static let maximumItemCount = 1_000

    let type: String
    let version: Int
    let group: GroupManifestGroup
    let items: [GroupManifestItem]

    private enum CodingKeys: String, CodingKey {
        case type
        case version
        case group
        case items
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        type = try container.decode(String.self, forKey: .type)
        version = try container.decode(Int.self, forKey: .version)
        group = try container.decode(GroupManifestGroup.self, forKey: .group)
        items = try container.decode([GroupManifestItem].self, forKey: .items)

        guard type == Self.expectedType else {
            throw GroupManifestValidationError.unexpectedType(type)
        }
        guard version == Self.supportedVersion else {
            throw GroupManifestValidationError.unsupportedVersion(version)
        }
        guard items.count <= Self.maximumItemCount else {
            throw GroupManifestValidationError.tooManyItems(items.count)
        }

        var itemIDs = Set<String>()
        var positions = Set<Int>()
        for (expectedPosition, item) in items.enumerated() {
            guard itemIDs.insert(item.id).inserted else {
                throw GroupManifestValidationError.duplicateItemID(item.id)
            }
            guard positions.insert(item.position).inserted else {
                throw GroupManifestValidationError.duplicatePosition(item.position)
            }
            guard item.position == expectedPosition else {
                throw GroupManifestValidationError.invalidPosition(
                    expected: expectedPosition,
                    actual: item.position
                )
            }
        }
    }
}

struct GroupManifestGroup: Decodable, Equatable, Sendable {
    let id: String
    let name: String
    let description: String?
    let slug: String
    let revision: Int

    private enum CodingKeys: String, CodingKey {
        case id
        case name
        case description
        case slug
        case revision
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try requireNormalizedText(container.decode(String.self, forKey: .id), field: "group.id")
        name = try requireNormalizedText(container.decode(String.self, forKey: .name), field: "group.name")
        description = try optionalNormalizedText(
            container.decodeIfPresent(String.self, forKey: .description),
            field: "group.description"
        )
        slug = try requireNormalizedText(container.decode(String.self, forKey: .slug), field: "group.slug")
        revision = try container.decode(Int.self, forKey: .revision)
        guard revision > 0 else {
            throw GroupManifestValidationError.invalidField("group.revision")
        }
    }
}

enum GroupManifestItemKind: String, Decodable, Equatable, Sendable {
    case catalog
    case github
    case synced
}

struct GroupManifestItem: Decodable, Equatable, Sendable {
    let id: String
    let kind: GroupManifestItemKind
    let position: Int
    let name: String
    let description: String?
    let note: String?
    let installability: GroupManifestInstallability

    private enum CodingKeys: String, CodingKey {
        case id
        case kind
        case position
        case name
        case description
        case note
        case installability
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try requireNormalizedText(container.decode(String.self, forKey: .id), field: "item.id")
        kind = try container.decode(GroupManifestItemKind.self, forKey: .kind)
        position = try container.decode(Int.self, forKey: .position)
        guard position >= 0 else {
            throw GroupManifestValidationError.invalidField("item.position")
        }
        name = try requireNormalizedText(container.decode(String.self, forKey: .name), field: "item.name")
        description = try optionalNormalizedText(
            container.decodeIfPresent(String.self, forKey: .description),
            field: "item.description"
        )
        note = try optionalNormalizedText(
            container.decodeIfPresent(String.self, forKey: .note),
            field: "item.note"
        )
        installability = try container.decode(GroupManifestInstallability.self, forKey: .installability)
    }
}

enum GroupManifestMetadataOnlyReason: String, Decodable, CaseIterable, Equatable, Sendable {
    case releaseUnavailable = "release_unavailable"
    case sourceUnavailable = "source_unavailable"
    case sourceMismatch = "source_mismatch"
    case incompleteRelease = "incomplete_release"
    case invalidRelease = "invalid_release"
    case releaseSourceMismatch = "release_source_mismatch"
    case syncedMissing = "synced_missing"
    case syncedLocalOnly = "synced_local_only"
    case syncedAmbiguous = "synced_ambiguous"
    case syncedUnresolved = "synced_unresolved"
}

enum GroupManifestInstallability: Decodable, Equatable, Sendable {
    case installable(source: GroupManifestSource, release: GroupManifestRelease)
    case metadataOnly(reason: GroupManifestMetadataOnlyReason)

    private enum Status: String, Decodable {
        case installable
        case metadataOnly = "metadata_only"
    }

    private enum CodingKeys: String, CodingKey {
        case status
        case source
        case release
        case reason
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        switch try container.decode(Status.self, forKey: .status) {
        case .installable:
            self = .installable(
                source: try container.decode(GroupManifestSource.self, forKey: .source),
                release: try container.decode(GroupManifestRelease.self, forKey: .release)
            )
        case .metadataOnly:
            self = .metadataOnly(
                reason: try container.decode(GroupManifestMetadataOnlyReason.self, forKey: .reason)
            )
        }
    }
}

enum GroupManifestSource: Decodable, Equatable, Sendable {
    case catalog(id: String, catalogSkillID: String, normalizedRoot: String)
    case publicGitHub(id: String, repositoryID: String, repositorySlug: String, normalizedRoot: String)
    case privateGitHub(id: String)

    private enum Kind: String, Decodable {
        case catalog
        case publicGitHub = "public_github"
        case privateGitHub = "private_github"
    }

    private enum CodingKeys: String, CodingKey {
        case id
        case kind
        case catalogSkillID = "catalogSkillId"
        case normalizedRoot
        case repositoryID = "repositoryId"
        case repositorySlug
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let id = try requireNormalizedText(container.decode(String.self, forKey: .id), field: "source.id")
        switch try container.decode(Kind.self, forKey: .kind) {
        case .catalog:
            self = .catalog(
                id: id,
                catalogSkillID: try requireNormalizedText(
                    container.decode(String.self, forKey: .catalogSkillID),
                    field: "source.catalogSkillId"
                ),
                normalizedRoot: try requireNormalizedText(
                    container.decode(String.self, forKey: .normalizedRoot),
                    field: "source.normalizedRoot"
                )
            )
        case .publicGitHub:
            let repositoryID = try requireNormalizedText(
                container.decode(String.self, forKey: .repositoryID),
                field: "source.repositoryId"
            )
            guard repositoryID.allSatisfy(\.isNumber) else {
                throw GroupManifestValidationError.invalidField("source.repositoryId")
            }
            let repositorySlug = try requireNormalizedText(
                container.decode(String.self, forKey: .repositorySlug),
                field: "source.repositorySlug"
            )
            guard isRepositorySlug(repositorySlug) else {
                throw GroupManifestValidationError.invalidField("source.repositorySlug")
            }
            self = .publicGitHub(
                id: id,
                repositoryID: repositoryID,
                repositorySlug: repositorySlug,
                normalizedRoot: try requireNormalizedText(
                    container.decode(String.self, forKey: .normalizedRoot),
                    field: "source.normalizedRoot"
                )
            )
        case .privateGitHub:
            self = .privateGitHub(id: id)
        }
    }
}

struct GroupManifestRelease: Decodable, Equatable, Sendable {
    let id: String
    let commitSha: String
    let treeSha: String
    let skillMdSha: String

    var coordinates: SkillPackageCoordinates {
        SkillPackageCoordinates(
            commitSha: commitSha,
            treeSha: treeSha,
            skillMdSha: skillMdSha
        )
    }

    private enum CodingKeys: String, CodingKey {
        case id
        case commitSha
        case treeSha
        case skillMdSha
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try requireNormalizedText(container.decode(String.self, forKey: .id), field: "release.id")
        commitSha = try requireGitSha(container.decode(String.self, forKey: .commitSha), field: "release.commitSha")
        treeSha = try requireGitSha(container.decode(String.self, forKey: .treeSha), field: "release.treeSha")
        skillMdSha = try requireGitSha(container.decode(String.self, forKey: .skillMdSha), field: "release.skillMdSha")
    }
}

enum DeviceGroupManifestRouteError: Error, Equatable, Sendable {
    case invalidHandle
    case invalidGroupSlug
}

struct DeviceGroupManifestRoute: Equatable, Sendable {
    let handle: String
    let groupSlug: String

    init(handle: String, groupSlug: String) throws {
        let normalizedHandle = handle.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let normalizedSlug = groupSlug.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard isRouteSegment(normalizedHandle) else {
            throw DeviceGroupManifestRouteError.invalidHandle
        }
        guard isRouteSegment(normalizedSlug) else {
            throw DeviceGroupManifestRouteError.invalidGroupSlug
        }
        self.handle = normalizedHandle
        self.groupSlug = normalizedSlug
    }
}

private func requireNormalizedText(_ value: String, field: String) throws -> String {
    let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty, trimmed == value else {
        throw GroupManifestValidationError.invalidField(field)
    }
    return value
}

private func optionalNormalizedText(_ value: String?, field: String) throws -> String? {
    guard let value else { return nil }
    return try requireNormalizedText(value, field: field)
}

private func requireGitSha(_ value: String, field: String) throws -> String {
    guard value.utf8.count == 40, value.utf8.allSatisfy({ byte in
        (48...57).contains(byte) || (65...70).contains(byte) || (97...102).contains(byte)
    }) else {
        throw GroupManifestValidationError.invalidField(field)
    }
    return value.lowercased()
}

private func isRepositorySlug(_ value: String) -> Bool {
    let parts = value.split(separator: "/", omittingEmptySubsequences: false)
    return parts.count == 2 && parts.allSatisfy { !$0.isEmpty }
}

private func isRouteSegment(_ value: String) -> Bool {
    guard !value.isEmpty, value.utf8.count <= 80 else { return false }
    return value.split(separator: "-", omittingEmptySubsequences: false).allSatisfy { part in
        !part.isEmpty && part.utf8.allSatisfy { byte in
            (48...57).contains(byte) || (97...122).contains(byte)
        }
    }
}
